File: backend/src/experiential/registry.py
"""
Public registry API for pedagogical methods, used by the experiential routes.

Importing this module triggers method discovery (via `methods/__init__.py`),
which imports every method file and runs its `method(...)` registration.
"""
from typing import Any, Dict, List, Optional

from src.experiential.methods import base
from src.experiential import methods  # noqa: F401  side-effect: discovers all methods

# The method used when the client sends an unknown / missing id.
DEFAULT_METHOD_ID = 'generic'


def list_methods() -> List[Dict[str, str]]:
    """Method metadata for the professor's picker (no system prompt)."""
    return [
        {
            'id': m.id,
            'label': m.label,
            'description': m.description,
            'prompt_hint': m.prompt_hint,
        }
        for m in base.METHODS.values()
    ]


def get_method(method_id: str) -> Optional[base.Method]:
    return base.METHODS.get(method_id)


def get_system_prompt(method_id: str) -> str:
    """Resolve a method id to its spine prompt, falling back to the default."""
    m = base.METHODS.get(method_id) or base.METHODS.get(DEFAULT_METHOD_ID)
    return m.system_prompt if m else ''


def get_schema(method_id: str) -> str:
    """The frontend method id (validator + player) a generated lab should use.

    Stamped onto the lab as config.method so the player page mounts the right
    pedagogy. Falls back to the default method's schema.
    """
    m = base.METHODS.get(method_id) or base.METHODS.get(DEFAULT_METHOD_ID)
    return (m.schema if m else None) or 'predict-reveal'

File: frontend/src/methods/registry.js
// Registry of experiential pedagogical methods.
//
// A "method" is one teaching pedagogy with its OWN schema (validator) and its
// OWN player. A lab records which one it is via `config.method`; the player page
// reads that and mounts the matching validator + Runner.
//
// Two ways a method registers:
//   1. Drop a SUBFOLDER here that default-exports { id, label, validate, Runner }
//      — it is auto-discovered below, no edits to this file.
//   2. registerMethod({ ... }) called at import time. Used by the built-in
//      `predict-reveal`, whose large player still lives in
//      pages/ExperientialPage.jsx; it can be relocated into a subfolder later
//      with no change to this seam.
//
// To add a brand-new pedagogy end-to-end:
//   - backend: src/experiential/methods/<id>.py  (the generation prompt; set its
//     `schema='<id>'` so generated labs are stamped config.method = '<id>')
//   - frontend: src/methods/<id>/index.{js,jsx}  default-exporting the descriptor
//     { id:'<id>', label, validate, Runner }

export const DEFAULT_METHOD_ID = 'predict-reveal';

const REGISTRY = {};

/** Register a method descriptor: { id, label, validate, Runner }. */
export function registerMethod(def) {
  if (def && def.id) REGISTRY[def.id] = def;
}

/** The descriptor for an id, or null if unknown. */
export function getMethod(id) {
  return REGISTRY[id] || null;
}

/** Lightweight list for pickers. */
export function listMethods() {
  return Object.values(REGISTRY).map(({ id, label }) => ({ id, label }));
}

// Auto-discover self-contained method subfolders (Vite glob, eager so they
// register at module load). A subfolder with no index file is simply ignored.
const modules = import.meta.glob('./*/index.{js,jsx}', { eager: true });
for (const mod of Object.values(modules)) {
  if (mod && mod.default) registerMethod(mod.default);
}
